### PR TITLE
[WIP] Port to Data.ByteString.Builder

### PR DESCRIPTION
**Pull request by bgamari**
This is work-in-progress porting the package away from `blaze-builder` in favor of `bytestring`'s own `Builder` which should be considerably faster.

This of course will require an API break and therefore will likely want to be a major version bump.

### CODE REVIEW

**Comment by aristidb** [comment]
We support blaze-builder 0.4, which is just a wrapper around bytestring-builder. Does this also bring performance gains compared to that?

File: Network/HTTP/Types/URI.hs [hunk 1]
```diff
@@ -252,18 +253,18 @@ urlDecode replacePlus z = fst $ B.unfoldrN (B.length z) go z
 -- Huge thanks to Jeremy Shaw who created the original implementation of this
 -- function in web-routes and did such thorough research to determine all
 -- correct escaping procedures.
-encodePathSegments :: [Text] -> Blaze.Builder
+encodePathSegments :: [Text] -> BSB.Builder
 encodePathSegments [] = mempty
 encodePathSegments (x:xs) =
-    Blaze.copyByteString "/"
+    BSB.byteString "/"
```
Comment (aristidb) [inline Network/HTTP/Types/URI.hs]: Why not BSB.char8 '/' here as well?:)

File: http-types.cabal [hunk 2]
```diff
@@ -35,7 +35,6 @@ Library
                        bytestring >=0.9.1.5 && <0.11,
                        array >=0.2 && <0.6,
                        case-insensitive >=0.2 && <1.3,
-                       blaze-builder >= 0.2.1.4 && < 0.5,
```
Comment (aristidb) [inline http-types.cabal]: Shouldn't there be an explicit dependency on bytestring-builder?

**Comment by bgamari** [comment]
I'm currently in the process of optimizing this. This work arose out of  discussion ensuing from the [`buffer-builder`](https://www.reddit.com/r/haskell/comments/2wwbuj/announcing_bufferbuilder_encode_json_25x_faster/) announcement earlier this week. Chad reports that much of his runtime is spent in `http-type`'s `urlEncode` implementation. I have experience tracking down performance issues with bytestring's `Builder` implementation so the first step in tracking down the issue was to eliminate the unnecessary trip through blaze. Moreover, I suspect that blaze is missing some key `INLINE` pragmas which will severely limit the compiler's ability to optimize. Given that `ByteString.Builder` is the canonical builder abstraction at this point, it seems reasonable to move `http-types` in this direction.

File: http-types.cabal (hunk 2 above)
Comment (bgamari) [inline http-types.cabal]: No, `Builder` is now included in `bytestring` as `Data.ByteString.Builder`. That being said, I do need to update the lower bound on `bytestring`.

File: Network/HTTP/Types/URI.hs (hunk 1 above)
Comment (bgamari) [inline Network/HTTP/Types/URI.hs]: Lovely point.

**Comment by aristidb** [comment]
Yes, if this improves performance we can go this way.

**Comment by jkarni** [comment]
Any updates on this?

**Comment by bgamari** [comment]
At this point I should benchmark this again to find out what the performance impact looks like. When I left this I had realized that bytestring's builder implementation was missing a rather major optimization opportunity. Now the fix is upstream, I'll have a look at this later this weekend.

**Comment by aristidb** [comment]
I think the new blaze-builder now uses bytestring builder, so I don't
expect major performance differences from using it directly? Still,
probably worth a try.

Ben Gamari <notifications@github.com> schrieb am Sa., 11. Juli 2015 um
10:45 Uhr:

**Comment by bgamari** [comment]
@aristidb that was the case when I wrote this as well. However, I seem to recall that `blaze-builder` was missing some crucial `INLINE` pragmas which would inhibit GHC's ability to fuse.

**Comment by 3noch** [comment]
@aristidb Updates?

**Comment by aristidb** [comment]
Fixed in 456bb08bc946322370993f9c6bfb12729896b364

**Pull request closed without merging**